areader/reader.go

```diff
@@ -599,18 +599,16 @@ func (ar *Reader) ReadArtifactHeaders() error {
 		return err
 	}
 
-	for no, us := range ar.updateStorers {
-		err = us.Initialize(ar.hInfo, ar.augmentedhInfo, ar.installers[no])
```

Comment:
Maybe I should have added both commit messages to the description:

> Delay constructing UpdateStorers until they are requested. This allows us to insert calls to `SetUpdateStorerProducer` in between calls to `ReadArtifactHeaders` and `ReadArtifactData`, which in turn allows us to store files for payload types we don't know about in advance. This will be used to extract data files from arbitrary artifacts.

So it allows us to store payload files from unknown payload types. Previously we couldn't because we needed to call `SetUpdateStorerProducer` before calling `ReadArtifactHeaders`, but we couldn't do that before knowing which type to use, which we get from `ReadArtifactHeaders`. So it was a chicken and egg problem before.

Reply:
I see, nice!